include/openthread/history_tracker.h

```diff
@@ -0,0 +1,211 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *
+ * @brief
+ *   This file includes the OpenThread API for History Tracker module.
+ */
+
+#ifndef OPENTHREAD_HISTORY_TRACKER_H_
+#define OPENTHREAD_HISTORY_TRACKER_H_
+
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/thread.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief
+ *   This header defines the public API for History Tracker.
+ *
+ *   History Tracker module records history of different events (e.g. RX and TX messages or network info changes, etc.)
+ *   as the Thread network operates. All tracked entries are timestamped.
+ *
+ *   The functions in this module are available when `OPENTHREAD_CONFIG_HISTOR_TRACKER_ENABLE` is enabled.
+ *
+ */
+
+/**
+ * This constant specifies the maximum age of entries which is 49 days (in msec).
+ *
+ * Entries older than the max age will give this value as their age.
+ *
+ */
+#define OT_HISTORY_TRACKER_MAX_AGE (49 * 24 * 60 * 60 * 1000u)
```

Comment:
I don't think it will be useful to make it configurable.

Users can not increase it beyond 49 days, and if they want a shorter interval, then they can simply ignore entries that have longer age.

This value is coming from our design and the fact that we use `uint32_t` value (as `TimeMilli`). `2^32` msec is about 49 days and 17 hours. So we track entries up to 49 days and every 16 hours check for entries that are about to reach the max (mark then as old).